[Guo Wei] iP

guowei42 opened the pull request:
# Deku Bot
> If you have an idea that you genuinely think is good, don't let some idiot talk you out of it. -- Stan Lee ([source](https://seancroxton.com/quote-of-the-day/742/#:~:text=Stan%20Lee%20%E2%80%9CIf%20You%20Have,Talk%20You%20Out%20of%20It.%E2%80%9D))

Use Deku to keep track of your tasks easily. It's
- text-based
- _so easy_ that even a ~10~ 6 years old can learn it

And it's literally **FREE**

All you need to do is:
1. Download it [here](https://github.com/guowei42/ip/releases/tag/A-Jar)
2. Double-click the `ip.jar` file
3. Add your tasks
4. Deku will manage your tasks for you😁

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Searching by date and word
- [ ] GUI (WIP)
---
Here is the `main` method:
``` java
public static void main(String[] args) throws IOException {
        Path directoryPath = Paths.get(System.getProperty("user.dir"), "data");
        Path filePath = directoryPath.resolve("save.txt");
        new Deku(directoryPath, filePath).start();
}
```

Commit:
Add Gradle support

guowei42 committed:
renamed default file

guowei42 committed:
add greeting, echo, exit

guowei42 committed:
added Task class

guowei42 committed:
added new BotList class

guowei42 committed:
added basic add and list feature, docs for methods

guowei42 committed:
WIP marking and unmarking

guowei42 committed:
added marking tasks as done or not done

guowei42 committed:
created new file and chmod

guowei42 committed:
add classes for events, deadlines features

guowei42 committed:
add todo, deadline and event command with date/time parsing

guowei42 committed:
add some basic text ui testing

guowei42 committed:
add error handling

guowei42 committed:
added additional test cases for error handling

guowei42 committed:
bug fix for file access issue

guowei42 committed:
add delete and updated test cases

guowei42 committed:
fix ghost file bug

guowei42 committed:
fix test file naming

guowei42 committed:
add enums, streamline command call

guowei42 committed:
fix bug for test case failure

guowei42 committed:
WIP save feature

guowei42 committed:
add save function after each list alteration

guowei42 committed:
fix javdoc format mistake

guowei42 committed:
initial commit

guowei42 committed:
add date and time parsing, searching via date and updated test cas

guowei42 committed:
fix test case failure

guowei42 committed:
add load function to Level-7

guowei42 committed:
rebasing files

guowei42 committed:
remove data file in runtest after each run

guowei42 committed:
organize classes into packages

guowei42 committed:
package mistake

guowei42 committed:
fix package duplicate

guowei42 committed:
A-Packages

guowei42 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

guowei42 committed:
add gradle

guowei42 committed:
final fix package

guowei42 committed:
add gradle support with building, running Deku class and checkstyle

guowei42 committed:
add some junit tests

guowei42 committed:
jar build

guowei42 committed:
improve coding standard and checkstyle violation

guowei42 committed:
add find word

guowei42 committed:
merge branches A-JavaDoc, A-CodingStandard, Level-9, resolve conflicts

KSHan29 commented on src/main/java/deku/BotList.java:
```diff
@@ -0,0 +1,127 @@
+               + getNoTasks();
+    }
+
+    String find_date(String date) throws DekuExceptions {
```
Perhaps the method name should be in camelCase instead?

KSHan29 commented on src/main/java/deku/BotList.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    String find_date(String date) throws DekuExceptions {
+        List<Task> containsDate = new ArrayList<>();
```
This might sound like a boolean type variable. Perhaps the name could be structured differently, eg DatesOfTasks

KSHan29 commented on src/main/java/deku/Deku.java:
```diff
@@ -0,0 +1,46 @@
+    private void start() {
+        Scanner scanner = new Scanner(System.in);
+        System.out.println(ui.introduction());
+        boolean active = true;
```
Perhaps could be phrased more like a boolean variable? eg isActive

KSHan29 commented on src/main/java/deku/InputParser.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public String parseTask(List<String> task) {
+        String date = "";
+        String misc = "";
```
Maybe this could be more descriptive?
If this is the part of user input which varies from command to command, maybe could name it as "commandInfo" so easier to see what the variable is for?

KSHan29 commented on src/main/java/deku/task/Task.java:
```diff
@@ -0,0 +1,116 @@
+    private final String icon;
+    private Boolean completionStatus;
+    private String completionIcon;
+    private final InputParser parser = new InputParser();
```
Perhaps the final variables should be in all capital letters to represent constants?

KSHan29 commented on src/test/java/deku/DekuExceptionTest.java:
```diff
@@ -0,0 +1,25 @@
+public class DekuExceptionTest {
+    @Test
+    public void ExceptionTest_emptyException_String() {
+        assertEquals("AUUUUUGH! ",
```
Good format but perhaps change String to the expected behaviour such that the naming follows recommended guideline:
featureUnderTest_testScenario_expectedBehavior();

KSHan29 reviewed (commented):
Overall I think good job!
There are some areas with similar mistakes:
1. Using underscore for naming methods when they should be in camelCase
2. Using camelCase for naming constants when they should be in all uppercase separated by underscore

guowei42 committed:
add JavaFx support and credit

guowei42 committed:
add basic design, remove text test case

guowei42 committed:
add basic GUI

autumn-sonata commented on src/main/java/deku/DekuExceptions.java:
```diff
@@ -0,0 +1,26 @@
+    public DekuExceptions(String errorMsg) {
+        super("AUUUUUGH! "
+              + errorMsg);
+        message = errorMsg;
+    }
+
+    @Override
+    public String toString() {
+        return "AUUUUUGH! "
+               + message;
```
I think that this is redundant. You should use the `getMessage()` method inbuilt within the `Exception` class itself.

autumn-sonata commented on src/main/java/deku/InputParser.java:
```diff
@@ -0,0 +1,158 @@
+        return output.substring(0, output.length() - 1);
+    }
+
+    public LocalDate getDate() {
+        return this.date;
+    }
+
+    public LocalTime getTime() {
+        return this.time;
+    }
```
Javadocs are needed for these 2 methods 👍

autumn-sonata commented on src/main/java/deku/UI.java:
```diff
@@ -0,0 +1,11 @@
+package deku;
+
+class UI {
```
UI should be named as Ui instead.
> Abbreviations and acronyms should not be uppercase when used as a part of a name.
```suggestion
class Ui {
```

autumn-sonata commented on src/main/java/deku/task/Task.java:
```diff
@@ -0,0 +1,116 @@
+
+    Task(List<String> task, String taskName, String icon) throws DekuExceptions {
+        if (task.size() == 0) {
+            throw new DekuExceptions("The description of a " + taskName + " cannot be empty.");
```
>In general, try to avoid any magic literals.

Consider making a constant variable for this instead or create an `EmptyDescriptionException` class that inherits `DekuExceptions` for this.

autumn-sonata commented on src/main/java/deku/BotList.java:
```diff
@@ -0,0 +1,128 @@
+        return outputList("Here are the tasks with the same date:", containsDate);
+    }
+
+    String find_word(String word) throws DekuExceptions {
```
>Names representing methods must be verbs and written in camelCase.

Please use the camelCase method of naming your method.

autumn-sonata commented on src/main/java/deku/DekuExceptions.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Custom exception class for bot when faced with unexpected results
+ */
+public class DekuExceptions extends Exception {
```
This class should be `DekuException`, not `DekuExceptions`.

autumn-sonata commented on src/main/java/deku/InputParser.java:
```diff
@@ -0,0 +1,158 @@
+                isDateTime = true;
+                output += "(" + top.substring(1) + ": ";
+                index++;
+                continue;
+            }
+            if (isDateTime) {
+                if (date.equals("")) {
+                    date = top;
+                } else {
+                    time = top;
+                }
+                misc += top + " ";
+            } else {
+                output += top + " ";
+            }
+            index++;
+        }
+        if (!misc.equals("")) {
+            misc = misc.substring(0, misc.length() - 1);
+        }
+        if (isDateTime) {
+            parseDate(date);
+            if (this.date == null) {
+                return output + misc + ")";
+            }
+            output += this.date.format(DateTimeFormatter.ofPattern("d MMM yyyy"));
+            if (time.equals("")) {
+                return output + ")";
+            }
+            parseTime(time);
+            if (this.time == null) {
+                return output + " " + time + ")";
+            }
+            return output
+                   + " "
+                   + this.time.format(DateTimeFormatter.ofPattern("HH:mm"))
+                   + ")";
+        }
+        return output.substring(0, output.length() - 1);
+    }
```
> Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

Maybe you can parse your method into several methods instead. This will help improve readability of the code. 👍

autumn-sonata reviewed (commented):
Good work overall! However, please be more mindful of the code quality when writing code.

guowei42 committed:
Add assertions to fetching Tasks by index.

Tasks index should only take in positive numbers, excluding 0.
Otherwise, trying to fetch outside the array might cause
unexpected behaviours.

guowei42 committed:
update assert messages

guowei42 committed:
Restructure code to adhere to better code quality.

Add SLAP to InputParser.parseTask class and reduce its LOC. Remove
pyramid layered conditional statements. Remove unused class
Ui.java. Remove long logical statements.

Add extra logic for exiting the program when user inputs bye.

guowei42 committed:
Merge pull request #1 from guowei42/branch-A-CodeQuality

Branch a-code quality

guowei42 committed:
change permission

guowei42 committed:
resolve conflicts

guowei42 committed:
Merge pull request #2 from guowei42/branch-A-Assertions

Branch a assertions

guowei42 committed:
test CI integration

guowei42 committed:
fix checkstyle violations

guowei42 committed:
Merge pull request #3 from guowei42/branch-A-CI

test CI integration

guowei42 committed:
add undo feature.

Users can make use of the keyword undo to get rid of the last
command input. This allows the user to return to the last state.
Just in case they made a mistake and wants to quickly return back.

guowei42 committed:
Merge pull request #4 from guowei42/branch-BCD-Extension

add undo feature.

guowei42 committed:
fix file not saved directly after undo command issue

guowei42 committed:
small user guide update

guowei42 committed:
add rounded corner to profile pictures, add background image to panes

guowei42 committed:
fix checkstyle

guowei42 committed:
resolve issues, add assert check

guowei42 committed:
fix bug when reading save file, tasks are not separated correctly based on space character

guowei42 committed:
update user guide

guowei42 committed:
add more exception tests and Ui sample image

guowei42 committed:
update user guide

guowei42 committed:
update user guide formatting

guowei42 committed:
update user guide formatting

guowei42 committed:
update credit

guowei42 committed:
change .jpg to .png

guowei42 committed:
fix truncate text after a certain length, update test cases

guowei42 committed:
fix deadline command not outputing as intended, fix find_date find_word bug

guowei42 committed:
add credit

guowei42 committed:
bug fixes

guowei42 committed:
checkstyle

guowei42 committed:
remove sysout
